### PR TITLE
feat(ui): shop item pane autoscrolling and handling vanished items

### PR DESCRIPTION
**Pull request by flowers9**
**Feature:**

## Feature Details
This makes the key-based movement in the main pane of the ShopPanel work properly - autoscrolling happens when moving to something off-screen (instead of keeping the current selection Y fixed and scrolling the screen every time you go up or down a row).  There is no longer a blank entry between the last and first items.

When items vanish (from being sold, stored, bought into cargo, uninstalled, or from the various toggles, from changing the current ship, or from collapsing a category) the current selection advances to the next displayed item (instead of continuing to show the old non-displayed item).  In most cases, the main pane will also scroll to show the newly selected item.

Removed a number of unused variables and some code that never did anything because those variables never changed value.

## Testing Done
Lots of scrolling in the item pane (left, right, up, down, around corners, etc.).  Making the selected item vanish in a number of ways - toggles, ship selection, buying/selling.

## Performance Impact
N/A

## Additional Notes
I would have done the same to the ship sidebar, but I don't want to mess with that until #9069 or #9074 goes through, since it changes all that code.

This code supercedes #9072 and fixes the problem I had with that approach, as the code no longer cares about the relative order of the clickzones between main and side panes.

Thanks to Koranir for the smooth scrolling changes!

### CODE REVIEW

**Comment by Koranir** [comment]
While we're improving the Shop panel: https://github.com/flowers9/endless-sky/pull/2

**Comment by flowers9** [comment]
Not sure why that test failed.  It appears to be completely unrelated to the code changes, and passed when I ran it locally (ubuntu 22.04).

**Comment by TomGoodIdea** [comment]
This should resolve #3088.

**Comment by flowers9** [comment]
Turns out ships are stocked by variant name, but everything else uses truemodelname.  Yay.

File: source/ShopPanel.cpp [hunk 1]
```diff
@@ -1142,194 +1128,272 @@ void ShopPanel::SideSelect(Ship *ship)
...
 void ShopPanel::MainUp()
 {
-	vector<Zone>::const_iterator start = MainStart();
-	if(start == zones.end())
+	if(zones.empty())
 		return;
 
 	vector<Zone>::const_iterator it = Selected();
-	// Special case: nothing is selected. Go to the last item.
+	// Special case: nothing is selected.  Start from the first item.
```
[the remark this answers is not in the document]
Comment (flowers9) [reply source/ShopPanel.cpp:1214]: Left/right movement wrap around lines, but up/down movement tries to preserve the position in the line (subject to the line being moved too being too short to do so).

So, when nothing is selected, I treat left/up movement as from the first item (selecting the last item/first item on last row, respectively), whereas right/down both select the first item.

[the remark this answers is not in the document]
Comment (flowers9) [reply source/ShopPanel.cpp:1214]: Essentially, up/down while nothing is selected are treated as being in the first column.  It's like the "nothing" entry is a single row with one item, situated between the last and first rows.

**Review by quyykk — APPROVED** [review]
Thanks!

File: source/ShopPanel.h [hunk 2]
```diff
@@ -202,15 +200,20 @@ class ShopPanel : public Panel {
 	bool SetScrollToBottom();
 	void SideSelect(int count);
 	void SideSelect(Ship *ship);
+	void MainAutoScroll(const std::vector<Zone>::const_iterator &selected);
 	void MainLeft();
 	void MainRight();
 	void MainUp();
 	void MainDown();
+	void CategoryAdvance(const std::string &);
```
Comment (Hurleveur) [inline source/ShopPanel.h:208]:
```suggestion
	void CategoryAdvance(const std::string &category);
```
we generally write the name of the variable in both places

Comment (flowers9) [reply source/ShopPanel.h]: Yeah, don't know how I missed that one.

**Review by Hurleveur — APPROVED** [review]
looks good codewise, and it worked wonders last time I tried it, good work, I hope we can even push it further in a follow up PR

**Pull request merged**